File: API/1661_Symr_New_Bar/CS/SymrNewBarStrategy.cs
using System;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Strategy that logs when new bars start for multiple timeframes.
/// It tracks predefined periods and notifies when a new bar appears.
/// </summary>
public class SymrNewBarStrategy : Strategy
{
	private readonly StrategyParam<DataType> _candleType;

	// Supported periods in ascending order.
	private static readonly TimeSpan[] _periods =
	{
		TimeSpan.FromMinutes(1),
		TimeSpan.FromMinutes(5),
		TimeSpan.FromMinutes(15),
		TimeSpan.FromMinutes(30),
		TimeSpan.FromHours(1),
		TimeSpan.FromHours(4),
		TimeSpan.FromDays(1),
		TimeSpan.FromMinutes(20),
		TimeSpan.FromMinutes(55),
	};

	// Last known open time for each period.
	private readonly DateTimeOffset[] _times = new DateTimeOffset[_periods.Length];

	private int _currentIndex;

	/// <summary>
	/// Base candle type used to detect new bars.
	/// </summary>
	public DataType CandleType
	{
		get => _candleType.Value;
		set => _candleType.Value = value;
	}

	/// <summary>
	/// Initializes parameters.
	/// </summary>
	public SymrNewBarStrategy()
	{
		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(1).TimeFrame())
			.SetDisplay("Candle Type", "Primary timeframe to monitor", "General");
	}

	/// <inheritdoc />
	public override System.Collections.Generic.IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return [(Security, CandleType)];
	}

	/// <inheritdoc />
	protected override void OnStarted(DateTimeOffset time)
	{
		base.OnStarted(time);

		// Determine index of the base timeframe.
		_currentIndex = Array.IndexOf(_periods, ((CandleTimeFrame)CandleType.Arg).TimeSpan);

		var subscription = SubscribeCandles(CandleType);
		subscription.Bind(ProcessCandle).Start();
	}

	private void ProcessCandle(ICandleMessage candle)
	{
		if (candle.State != CandleStates.Finished)
			return;

		var currentTime = candle.OpenTime;

		if (_times[_currentIndex] != currentTime)
		{
			_times[_currentIndex] = currentTime;
			OnBar(_periods[_currentIndex], currentTime);

			for (var i = _currentIndex + 1; i < _periods.Length; i++)
			{
				var period = _periods[i];
				var time0 = currentTime - TimeSpan.FromTicks(currentTime.Ticks % period.Ticks);

				if (_times[i] != time0)
				{
					_times[i] = time0;
					OnBar(period, time0);
				}
			}
		}

		OnTick();
	}

	private void OnTick()
	{
		// Called after each processed candle.
		// Placeholder for tick-based logic.
	}

	private void OnBar(TimeSpan period, DateTimeOffset time)
	{
		// Log new bar event for the specified period.
		LogInfo("New bar at {time} for period {period}.");
	}
}

File: API/1661_Symr_New_Bar/PY/symr_new_bar_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import ExponentialMovingAverage
from StockSharp.Algo.Strategies import Strategy


class symr_new_bar_strategy(Strategy):
    def __init__(self):
        super(symr_new_bar_strategy, self).__init__()
        self._ema_length = self.Param("EmaLength", 20) \
            .SetDisplay("EMA Length", "EMA period for trend filter", "Indicators")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(4))) \
            .SetDisplay("Candle Type", "Type of candles", "General")
        self._prev_high = 0.0
        self._prev_low = 0.0
        self._has_prev = False

    @property
    def ema_length(self):
        return self._ema_length.Value

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(symr_new_bar_strategy, self).OnReseted()
        self._prev_high = 0.0
        self._prev_low = 0.0
        self._has_prev = False

    def OnStarted(self, time):
        super(symr_new_bar_strategy, self).OnStarted(time)
        ema = ExponentialMovingAverage()
        ema.Length = self.ema_length
        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(ema, self.on_process).Start()
        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawOwnTrades(area)

    def on_process(self, candle, ema_val):
        if candle.State != CandleStates.Finished:
            return
        close = candle.ClosePrice
        if not self._has_prev:
            self._prev_high = candle.HighPrice
            self._prev_low = candle.LowPrice
            self._has_prev = True
            return
        # Breakout above previous high with EMA confirmation
        if close > self._prev_high and close > ema_val and self.Position <= 0:
            if self.Position < 0:
                self.BuyMarket()
            self.BuyMarket()
        # Breakout below previous low with EMA confirmation
        elif close < self._prev_low and close < ema_val and self.Position >= 0:
            if self.Position > 0:
                self.SellMarket()
            self.SellMarket()
        self._prev_high = candle.HighPrice
        self._prev_low = candle.LowPrice

    def CreateClone(self):
        return symr_new_bar_strategy()
